test/boost/secondary_index_test.cc

```diff
@@ -1560,6 +1560,305 @@ SEASTAR_TEST_CASE(test_token_order) {
+            assert_that(msg).is_rows().with_rows(get_result_rows(1, 3));
+
+            // exclusive-exclusive range
+            msg = cquery_nofail(e, format("SELECT * FROM t WHERE v = 1 AND token(pk) > {} AND token(pk) < {} ALLOW FILTERING", testset_tokens[1], testset_tokens[4]).c_str());
+            assert_that(msg).is_rows().with_rows(get_result_rows(2, 3));
+
+            // inclusive-infinity range
+            msg = cquery_nofail(e, format("SELECT * FROM t WHERE v = 1 AND token(pk) >= {} ALLOW FILTERING", testset_tokens[3]).c_str());
+            assert_that(msg).is_rows().with_rows(get_result_rows(3, testset_pks.size() - 1));
+
+            // exclusive-infinity range
+            msg = cquery_nofail(e, format("SELECT * FROM t WHERE v = 1 AND token(pk) > {} ALLOW FILTERING", testset_tokens[3]).c_str());
+            assert_that(msg).is_rows().with_rows(get_result_rows(4, testset_pks.size() - 1));
+
+            // infinity-inclusive range
+            msg = cquery_nofail(e, format("SELECT * FROM t WHERE v = 1 AND token(pk) <= {} ALLOW FILTERING", testset_tokens[3]).c_str());
+            assert_that(msg).is_rows().with_rows(get_result_rows(0, 3));
+
+            // infinity-exclusive range
+            msg = cquery_nofail(e, format("SELECT * FROM t WHERE v = 1 AND token(pk) < {} ALLOW FILTERING", testset_tokens[3]).c_str());
+            assert_that(msg).is_rows().with_rows(get_result_rows(0, 2));
+
+            // negative-positive range
+            auto start_index =  3;
+            auto end_index = 5;
+            assert(testset_tokens[start_index] < 0);
+            assert(testset_tokens[end_index] > 0);
+
+            msg = cquery_nofail(e, format("SELECT * FROM t WHERE v = 1 AND token(pk) >= {} AND token(pk) <= {} ALLOW FILTERING", testset_tokens[start_index], testset_tokens[end_index]).c_str());
+            assert_that(msg).is_rows().with_rows(get_result_rows(start_index, end_index));
+            
+            // tokens smaller than 0
+            msg = cquery_nofail(e, "SELECT * FROM t WHERE v = 1 AND token(pk) < 0 ALLOW FILTERING");
+            assert_that(msg).is_rows().with_rows(get_result_rows(0, 4));
+
+            // tokens larger than 0
+            msg = cquery_nofail(e, "SELECT * FROM t WHERE v = 1 AND token(pk) > 0 ALLOW FILTERING");
+            assert_that(msg).is_rows().with_rows(get_result_rows(5, testset_pks.size() - 1));
+
+            // token equality
+            msg = cquery_nofail(e, format("SELECT * FROM t WHERE v = 1 AND token(pk) = {} ALLOW FILTERING", testset_tokens[3]).c_str());
+            assert_that(msg).is_rows().with_rows(get_result_rows(3, 3));
+
+            // empty range
+            msg = cquery_nofail(e, "SELECT * FROM t WHERE v = 1 AND token(pk) < 5 AND token(pk) > 100 ALLOW FILTERING");
+            assert_that(msg).is_rows().with_size(0);
+        };
+
+        eventually(do_tests);
```

Comment:
a comment not really related to this patch but maybe will interest you and @dekimir  - I don't like to write tests with eventually(). Every test failure (which are common during development) takes ages, while it retries the same test over and over - long after the index is built. In my new cql-pytest code I used a different approach: I build an index and then wait for the data to be available in this index. Only then I try the actual test I want to do - without any eventually(). If that test fails, it will fail immediately.

Reply:
Nice approach, but it doesn't cover all paths - view building and updating an already existing view exercise different code paths, so we need tests for both. If a view already exists and gets updated, I'm not aware of any other mechanisms than `eventually` to make sure that the updates already arrived. One can also poll the materialized view directly, but it's much less convenient to write and depends on implementation details.

Reply:
First, my suggestion doesn't need to cover **all** uses in all tests we ever wrote - it is aimed at tests like this one that I'm reviewing now - where you want to set up an index (or a view), and then run various tests on this index (or view).

Second, my suggestion **does** work even when you update a view during the test, provided the table is small enough for the approach to make sense. Quite simply, after updating the index/view, you need to again call the waiting function, which reads the whole table and then reads the whole table through the index/view - and compares the two and wait until they finally compare equal. Only when they finally compare equal, you can run the  actual check that you wanted to check, and you know that the index is already up to date and don't need an "eventually". Obviously, there is one case where this approach doesn't make much sense - if the only purpose of the eventually() was to verify that something written to the table is readable in the view, it doesn't make too much sense to read everything instead of just the one item you want to verify.